python/user/bot/user_bot.py

```diff
@@ -26,29 +45,82 @@ def main():
+        archive_state.message_heap = [Message(int(cdata['postedAt']), cid)
+                                      for cid, cdata in existing_messages.items()]
+
+        message_heap = archive_state.message_heap
+        heapq.heapify(message_heap)
+        logging.info(f"Message cache loaded.")
+
+        (_, settings_cdata) = await client.find_one(Chat.UserSettings, {'user': client.party})
+        archive_state.archive_after = (settings_cdata['archiveMessagesAfter'])
+
+        # initial scan
+        while len(message_heap) > 0:
+            top = message_heap[0]
+            if not expired(archive_state.archive_after, top.post_at):
+                logging.info("Auto-archiving caught up.")
+                break
+            logging.info(f"Auto-archiving {top.cid} posted at {top.post_at}")
+            await client.submit(exercise(top.cid, 'Archive'))
+            heapq.heappop(message_heap)
+
+        logging.info(f'started auto-archiving bot for party {party}')
+        while True:
+            try:
+                while len(message_heap) > 0 and expired(archive_state.archive_after,
+                                                        message_heap[0].post_at):
+                    top = message_heap[0]
+                    logging.info(f'archiving {Chat.Message}:{top.cid}'
+                                 f' expired after {archive_state.archive_after}s')
+                    await client.submit(exercise(top.cid, 'Archive'))
+                    heapq.heappop(message_heap)
+                    logging.info(f'{Chat.Message}:{top.cid} archived.')
+            except:
+                logging.error(f"Could not auto archive messages")
+            if len(message_heap) > 0:
+                logging.info(f'waiting for next message to archive: {message_heap[0].cid}')
+            await asyncio.sleep(bot_polling_sec)
 
     @client.ledger_created(Chat.ArchiveMessagesRequest)
     async def archive_stale_messages(event):
         logging.info(f'On {Chat.ArchiveMessagesRequest} created!')
         try:
             (_, settings_cdata) = await client.find_one(Chat.UserSettings, {'user': client.party})
-            time_thresh = datetime.datetime.now() - datetime.timedelta(
-                days=settings_cdata['archiveMessagesAfter'])
-            thresh_seconds = (time_thresh - EPOCH).total_seconds()
-            logging.info(f"time_thresh: {time_thresh}, thresh_seconds: {thresh_seconds}")
             user_messages = client.find_active(Chat.Message, {'sender': client.party})
             commands = [exercise(cid, 'Archive') for (cid, cdata) in user_messages.items() if
-                        int(cdata['postedAt']) < thresh_seconds]
+                        expired(settings_cdata['archiveMessagesAfter'], int(cdata['postedAt']))]
             logging.info(f"Will archive {len(commands)} message(s)")
             commands.append(exercise(event.cid, 'Archive'))
             await client.submit(commands)
```

Comment:
If I'm understanding the flow correctly, this list of commands will include every message that is stale at the time the archive request comes in.  For a a chat instance that's been around a long time, that list might grow to the point where the command cannot be done in a single submission/txn. 

I think you probably ought to work to submit commands in batches of finite size.

Reply:
Agree with this. but I feel dabl-chat isn't designed to handle large scale usage at all. otherwise lots of things can be optimized here... including a separate caching service, etc etc..